torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,1737 @@
+            # go out of scope, at which point it could otherwise be freed for
+            # further reuse by the main stream while the div/reduce_scatter/copy
+            # are underway in the post_backward stream. See:
+            # github.com/NVIDIA/apex/blob/master/apex/parallel/distributed.py
+            orig_grad_data.record_stream(self._streams["post_backward"])
+
+    def _post_reduction_hook(self, param: Parameter, reduced_grad: torch.Tensor) -> None:
+        """Hook to call on each param after the reduce-scatter."""
+        assert torch.cuda.current_stream() == self._streams["post_backward"]
+        assert param.grad is not None
+        self.assert_state(TrainingState.BACKWARD_POST)
+        param.grad.data = reduced_grad
+        if self.gradient_postdivide_factor > 1:
+            # Average grad by world_size for consistency with PyTorch DDP.
+            param.grad.data.div_(self.gradient_postdivide_factor)
+        # Cast grad to param's dtype (typically FP32). Note: we do this
+        # before the move_grads_to_cpu step so that this entire hook remains
+        # non-blocking. The downside is a bit more D2H transfer in that case.
+        if self.mixed_precision:
+            orig_param_grad_data = param.grad.data
+            param.grad.data = param.grad.data.to(dtype=param.data.dtype)
+            # Don't let this memory get reused until after the transfer.
+            orig_param_grad_data.record_stream(torch.cuda.current_stream())  # type: ignore[arg-type]
+        if hasattr(param, "_saved_grad_shard") and param._saved_grad_shard is not None:  # type: ignore[attr-defined]
+            assert (
+                param._saved_grad_shard.shape == param.grad.shape  # type: ignore[attr-defined]
+            ), f"{param._saved_grad_shard.shape} vs {param.grad.shape}"  # type: ignore[attr-defined]
+            param.grad.data += param._saved_grad_shard  # type: ignore[attr-defined]
+            delattr(param, "_saved_grad_shard")
+        # Optionally move gradients to CPU, typically used if one is running
+        # the optimizer on the CPU.
+        if self.move_grads_to_cpu:
+            param._cpu_grad.copy_(param.grad.data, non_blocking=True)  # type: ignore[attr-defined]
+            # Don't let this memory get reused until after the transfer.
+            param.grad.data.record_stream(torch.cuda.current_stream())  # type: ignore[arg-type]
+            param.grad.data = param._cpu_grad  # type: ignore[attr-defined]
+
+    def _queue_wait_for_post_backward(self) -> None:
+        """Try to queue a `wait_for_post_backward` callback.
+        Only called on root and only queue one callback at the beginning of
+        outer most backward.
+        """
+        assert self._is_root
+        if not self._post_backward_callback_queued:
+            self.assert_state([TrainingState.IDLE])
+            self._post_backward_callback_queued = True
+            Variable._execution_engine.queue_callback(self._wait_for_post_backward)
```

Comment:
do we need to worry about reentrant backward here?

Reply:
I made this change in fairscale repo. it is added here on root instance only to resolve reentrant backward issues. so far so good:)